src/pudl/transform/ferc1.py

```diff
@@ -6509,3 +6540,152 @@ def plants_steam_ferc1(
+    calc_components = pd.concat(calc_metas).pipe(
+        make_implict_dimensions_explict_in_calculaiton_components, table_dimensions
+    )
+    return calc_components
+
+
+def make_implict_dimensions_explict_in_calculaiton_components(
+    calculation_components: pd.DataFrame, table_dimensions: pd.DataFrame
+) -> pd.DataFrame:
+    """Make all of the implied dimensions in the calculation components explict.
+
+    The raw XBRL metadata's calculations, there is an implicit assumption that the
+    calculation is value within a set of shared axis/primary key values. We have
+    mannually added some calculation which span multiple tables where the calculation
+    is only applicable within a subset of these axis/primary key columns. For example,
+    the :ref:`utility_plant_summary_ferc1` table includes many different
+    ``utility_types``, but several fields in that table are calculable from fields in
+    the :ref:`plant_in_service` table - which only includes the ``utilty_type`` of
+    ``electric``. We have added some of these calculations which include this
+    increased detail regarding these dimension columns by adding these dimensions to
+    calculation components when they apply only to one dimension. This enabled us to
+    not need to specify every single dimension that was implied within the XBRL
+    metadata. But in order for the calculations to be applied uniformly, this function
+    makes all of those implict dimensions explict.
+
+    We are assuming that if a dimension was associted with a calculation component
+    before this treatment, then that is the only dimension that applies whereas if no
+    dimension was associated with a calculation component, that calculation component
+    should be applied across all relevant dimensions.
+
+    Relate all of the calculation components with all found dimensions from
+    :func:`table_dimensions_ferc1`, unless the calculation components already
+    contained information regarding dimensions.
+    """
+    dimensions = [
+        col
+        for col in table_dimensions.columns
+        if col not in ["table_name", "xbrl_factoid"]
+        and col in calculation_components.columns
+    ]
+    logger.info(f"Adding {dimensions=} into calculation component table.")
+    calc_comps_w_dims = calculation_components.copy()
+    # for each dimension, use split/apply/combine. when there are no dims explict in
+    # the calc components, merge in all of the dims.
+    for dim_col in dimensions:
+        null_dim_mask = calc_comps_w_dims[dim_col].isnull()
+        calc_comps_w_implied_dims = pd.merge(
+            calc_comps_w_dims[null_dim_mask].drop(columns=[dim_col]),
+            table_dimensions[["table_name", "xbrl_factoid", dim_col]]
+            .rename(
+                columns={
+                    "xbrl_factoid": "xbrl_factoid_calc",
+                    "table_name": "table_name_calc",
+                }
+            )
+            .drop_duplicates()
+            .dropna(),
```

Comment:
I'm finding it difficult to read this statement with the construction of the right dataframe inside the merge. Is this attempting to fill NA values found in the dimension columns?

Rather than using a merge here and iterating over all of the dimension columns, would it make sense to use `df1.fillna(df2)`?  If we give both dataframes the same index, we can fill NA values in `df1` using whatever values are found at the same location (row index + column name) in `df2`. Is this a `many-to-one` merge? I think this `fillna()` would broadcast values in the event that the row index didn't identify unique records in `df1`?

But maybe I'm still not understanding the structure of this table, or what it means to "relate all calculation components with all found dimensions..."

In the case of the `plant_in_service_ferc1` table, what happens?  That table has implied `utility_type=="electric"` and `plant_status=="in_service"` and there's a reference to it from a calculation in the `utility_plant_summary_ferc1` table that knows about the `utility_type` dimension. But not the `plant_status` dimension? Does it end up only being associated with `electric` but associated with all values of `plant_status`?

Reply:
awe i love the idea of using `fillna` but i don't think it supports one-to-many broadcasts

```
df1 = pd.DataFrame(
    [1, 2, pd.NA, 11, pd.NA, 14],
    index=["a"]*3 + ["b"]*3,
    columns=["data"]
)

df2 = pd.DataFrame(
    [3,4, 11, 12,13],
    index=["a"]*2 + ["b"]*3,
    columns=["data"]
)
df1.data.fillna(df2.data)
>>> ValueError: cannot reindex on an axis with duplicate labels
```

i did move the manipluation of `table_dimensions` out of the merge bc yes that was too confusing.

re your `plant_in_service_ferc1` example: for the `utlity_type` dimension, because it `electric` was added in the calc fixes, no additional `utility_type`'s get added. For `plant_status`, we are effectively checking for all of the `plant_status`'s that are observed in that `plant_in_service_ferc1` table and merging those on in order to broadcast the observed dimension. In this case, there are no `plant_status`'s found in `plant_in_service_ferc1` so nothing gets broadcast merged. But if two `plant_status`'s were observed in `plant_in_service_ferc1` it would be a one-to-many merge.